Product features

Tyler-Kuhn opened the pull request:
##  Add Products CRUD Feature

### Overview
This pull request adds full CRUD functionality for the **Products** feature.

### Features Added
-  Prisma integration for Product model
-  Service layer:
  - `createProduct`
  - `getAllProducts`
  - `getProductByName`
  - `updateProduct`
  - `deleteProduct`
-  Controller methods with error handling
-  Routes:
  - `POST /products/create`
  - `GET /products/allProducts`
  - `GET /products/product`
  - `PUT /products/update`
  - `DELETE /products/delete`
- Errors:
  - `Global error handler for normal errors`
  - `Global error handler for prisma errors`

### Notes
- Centralized error handling via `AppError` and `catchFunction`
- All routes tested and working via Postman
- updated route names to match best practices for nameing

Tyler-Kuhn committed:
Updated schcema for products,added prduct services

Tyler-Kuhn committed:
Added product productController

Tyler-Kuhn committed:
finished product routes

Tyler-Kuhn committed:
fixed files, chekced product routes

Tyler-Kuhn committed:
fixed files, chekced product routes

Tyler-Kuhn committed:
Fixed controller

worldwidewes33 commented on dist/products/productController.js:
```diff
@@ -1 +1,63 @@
 "use strict";
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
```
I would add your dist folder to your `.gitignore`. It isn't needed in your GitHub repo since I will only review your typescript.

worldwidewes33 commented on src/products/productController.ts:
```diff
@@ -0,0 +1,94 @@
+export const createAProduct = catchFunction(
+  async (req: Request, res: Response, next: NextFunction) => {
+    
+    const { name, price, description } = req.body;
```
You should add a check to confirm that name, price, and description were sent with the request.

worldwidewes33 commented on src/products/productController.ts:
```diff
@@ -0,0 +1,94 @@
+    if (!newProduct) {
+      const error = new AppError("Something wnet wrong", 500);
+      next(error);
+    }
```
This is an unnecessary check. The `createProduct` function already throws an error if the product is not created. Prisma actually guarantees to either return the created resource or throw an error.

worldwidewes33 commented on src/products/productController.ts:
```diff
@@ -0,0 +1,94 @@
+    if (!productList) {
+      const error = new AppError("Something went wrong", 500);
+      next(error);
+    }
```
Prisma guarantees a list will be returned or an error will be thrown. You only need to decide how an empty list should be handled.

worldwidewes33 commented on src/products/productController.ts:
```diff
@@ -0,0 +1,94 @@
+
+    res.status(200).json(product);
+  }
+);
```
If you are trying to get a single product. I suggest pulling that product by its ID, as the ID is unique and the name is not. if you want to pull all products with the same name, then I would look into creating a filter for your `getProducts` route.

worldwidewes33 commented on src/products/productController.ts:
```diff
@@ -0,0 +1,94 @@
+export const updateAProduct = catchFunction(
+  async (req: Request, res: Response, next: NextFunction) => {
+    const { id } = req.params;
+    const { name, price, description, reviews, ratings } = req.body;
```
Verify all of the necessary request body parameters were submitted.

worldwidewes33 commented on src/products/productController.ts:
```diff
@@ -0,0 +1,94 @@
+    if (!updatedProduct) {
+      const error = new AppError("Unable to update product", 500);
+      next(error);
+    }
```
Prisma guarantees the product will be updated or an exception will be thrown.

worldwidewes33 commented on src/products/productController.ts:
```diff
@@ -0,0 +1,94 @@
+      next(error);
+    }
+
+    res.status(201).json({Message: "Product deleted", deletedProduct})
```
I would use a 204: No Content or 200: Success status code. It depends on if your frontend will need to do something with a deleted product after the request is sent.

worldwidewes33 commented on src/products/productController.ts:
```diff
@@ -0,0 +1,94 @@
+    if(!deletedProduct){
+        const error = new AppError("Product not found", 404);
+      next(error);
+    }
```
If the product does not exist, Prisma will throw an exception. This code will never execute.

worldwidewes33 commented on src/products/productRoutes.ts:
```diff
@@ -0,0 +1,24 @@
+
+router.put("/update/:id", updateAProduct);
+
+router.delete("/delete/:id", deleteAProduct);
```
I would read this article on REST API URI best practices: https://restfulapi.net/resource-naming/. In general, you do not want verbs in your URIs. The action to be performed by each route is partially determined by the HTTP Method used. For example, a POST method to route `/products` means we want to create a product.

worldwidewes33 commented on src/products/productService.ts:
```diff
@@ -0,0 +1,60 @@
+
+  if (!newProduct) {
+    throw new AppError("Somthing went wrong", 500);
+  }
```
Guaranteed a product is created or an exception will occur.

worldwidewes33 commented on src/products/productService.ts:
```diff
@@ -0,0 +1,60 @@
+  if (!allProducts) {
+    throw new AppError("Something went wrong", 500);
+  }
+
```
`prisma.product.findMany()` will never return null. You should be handling those exception in your global error handler.

worldwidewes33 commented on src/products/productService.ts:
```diff
@@ -0,0 +1,60 @@
+  });
+
+  return updatedProduct;
+};
```
When calling this route, make sure to use the HTTP method PATCH. Since there are optional parameters, PATCH is more suitable than PUT. Typically, PUT is used for a complete replacement of the resource, meaning that all fields for the resource must be included in the request body.

worldwidewes33 commented on src/server.ts:
```diff
@@ -0,0 +1,44 @@
+app.use((err: any, req: Request, res: Response, next: NextFunction) => {
+  console.error(err);
+  res.status(500).json({ error: "Something went wrong" });
+});
```
Consider exporting this global error handler into its own file. You will need to provide more descriptive error messages for different types of errors because you are wrapping all of your middleware functions in the `catchFunction`. As your project progresses, you will find that 500 - Something went wrong error is not very helpful.

worldwidewes33 commented on src/users/userAuthServices.ts:
```diff
@@ -29,7 +26,19 @@ export const registerUser = async (
+
+  if (!user) {
+    throw new AppError("Invalid Email", 400);
+  }
```
Is this necessarily true? Could the email be a valid email but the user doesn't exist in the database?

worldwidewes33 commented on src/users/userAuthController.ts:
```diff
@@ -61,8 +60,8 @@ export const updateUser = catchFunction(
-    if(!updatedUser){
-      const error = new AppError("Could not update profile", 500)
+    if (!updatedUser) {
+      const error = new AppError("Could not update profile", 500);
```
Prisma will throw an exception if there was an error updating the user or if the user does not exist.

worldwidewes33 reviewed (changes requested):
Overall you have made great progress on your app. Most of my suggestions have to deal with error handling. 

1. I would decide where I want my error handling to occur. Does that happen in the controller or in the service function? It appears to are doing it in both. I would pick one and stick with it for code consistency. 
2. I want you to get clear on what you should expect the prisma client to return when you query the database. Here is the [Prisma API Reference](https://www.prisma.io/docs/orm/reference/prisma-client-reference). This will allow you to properly handle the common types of errors prisma will throw in your global error handler. 
3. I want you to read this [resource](https://restfulapi.net/resource-naming/) on REST API naming conventions. It should help you standardize your routes to conform with REST. Also I suggest getting clear on what HTTP methods you should be using for each reequest and what HTTP status code you should be sending. 

Let me know if you would like to book a call if these suggestions are not clear enough. I can show you some examples of what I am looking for.

Tyler-Kuhn committed:
Started fixes for redundancy and error handling

Tyler-Kuhn committed:
Updated Error handling and fixed route names

Tyler-Kuhn commented:
@worldwidewes33 have made the suggested update i believe, let me know what you think. 

added error handling for regular and prisma errors, and move the error handling to the controller layer, except there were a few for the userAuth that would give me errors if i removed them from the service layer because typescript was worried about certain variables being null or undefined. also fixed the route names

worldwidewes33 commented on src/products/productRoutes.ts:
```diff
@@ -0,0 +1,24 @@
+
+
+
+router.post("/product", createAProduct);
```
```suggestion
router.post("/products", createAProduct);
```

worldwidewes33 commented on src/server.ts:
```diff
@@ -0,0 +1,43 @@
+  res.send("API is running...");
+});
+
+app.use(prismaErrorHandler as ErrorRequestHandler);
+app.use(globalErrorHandler as unknown as ErrorRequestHandler); 
```
Since you are nexting to your global error handler, you can combine these two. into one handler. That may make it so you do not have to typecast.

worldwidewes33 commented on src/users/userAuthServices.ts:
```diff
@@ -14,17 +14,14 @@ export const registerUser = async (
+  await prisma.user.create({
     data: {
       name,
       email,
       password: hashedPassword,
     },
   });
 
-  if (!newUser) {
-    throw new AppError("Something went wrong", 500);
-  }
 
   const user: User | null = await prisma.user.findUnique({
     where: {
```
You can go back to the original implementation. The create function is guaranteed to not return null. It will throw an error if it is unable to create the user. No need to check of user is null. You do not need to make two requests to the database.

Tyler-Kuhn committed:
Update productRoutes.ts

Co-authored-by: Wesley Austin <131537918+worldwidewes33@users.noreply.github.com>
